modules/imgproc/src/drawing_text.cpp

```diff
@@ -0,0 +1,1541 @@
+//by amarullz from https://stackoverflow.com/questions/5423960/how-can-i-recognize-rtl-strings-in-c
+static bool isRightToLeft(unsigned c)
+{
+    if(c < 0x5BE)
+        return false;
+
+    return
+    ((c==0x05BE)||(c==0x05C0)||(c==0x05C3)||(c==0x05C6)||
+    ((c>=0x05D0)&&(c<=0x05F4))||
+    (c==0x0608)||(c==0x060B)||(c==0x060D)||
+    ((c>=0x061B)&&(c<=0x064A))||
+    ((c>=0x066D)&&(c<=0x066F))||
+    ((c>=0x0671)&&(c<=0x06D5))||
+    ((c>=0x06E5)&&(c<=0x06E6))||
+    ((c>=0x06EE)&&(c<=0x06EF))||
+    ((c>=0x06FA)&&(c<=0x0710))||
+    ((c>=0x0712)&&(c<=0x072F))||
+    ((c>=0x074D)&&(c<=0x07A5))||
+    ((c>=0x07B1)&&(c<=0x07EA))||
+    ((c>=0x07F4)&&(c<=0x07F5))||
+    ((c>=0x07FA)&&(c<=0x0815))||
+    (c==0x081A)||(c==0x0824)||(c==0x0828)||
+    ((c>=0x0830)&&(c<=0x0858))||
+    ((c>=0x085E)&&(c<=0x08AC))||
+    (c==0x200F)||(c==0xFB1D)||
+    ((c>=0xFB1F)&&(c<=0xFB28))||
+    ((c>=0xFB2A)&&(c<=0xFD3D))||
+    ((c>=0xFD50)&&(c<=0xFDFC))||
+    ((c>=0xFE70)&&(c<=0xFEFC))||
+    ((c>=0x10800)&&(c<=0x1091B))||
+    ((c>=0x10920)&&(c<=0x10A00))||
+    ((c>=0x10A10)&&(c<=0x10A33))||
+    ((c>=0x10A40)&&(c<=0x10B35))||
+    ((c>=0x10B40)&&(c<=0x10C48))||
+    ((c>=0x1EE00)&&(c<=0x1EEBB)));
+}
+#endif
+
+Point FontRenderEngine::putText_(
+    Mat& img, Size imgsize, const String& str_, Point org,
+    const uchar* color, FontFace& fontface, double size_,
+    double weight_, int flags, Range wrapRange, bool render, Rect* bbox_ )
+{
+    size_ = cvRound(size_*256)/256.;
+    bool bottom_left = (flags & PUT_TEXT_ORIGIN_BL) != 0;
+    int saved_weights[BUILTIN_FONTS_NUM+1]={0};
+    int size = cvRound(size_);
+    int weight = cvRound(weight_*65536);
+
+    if(fontface.getName().empty())
+        fontface.set("sans");
+    if(!fontface->ttface)
+        return org;
+
+    Point pen = org;
+    int i, j, len = (int)str_.size();
+    bool wrap = (flags & PUT_TEXT_WRAP) != 0;
+    int alignment = flags & PUT_TEXT_ALIGN_MASK;
+    int x0 = std::min(wrapRange.start, wrapRange.end);
+    int x1 = std::max(wrapRange.start, wrapRange.end);
+
+    if(x0 == 0 && x1 == 0)
+    {
+        if(alignment == PUT_TEXT_ALIGN_RIGHT)
+        {
+            x0 = 0;
+            x1 = org.x;
+        }
+        else
+        {
+            x0 = org.x;
+            x1 = imgsize.width > 0 ? imgsize.width : INT_MAX;
+        }
+    }
+
+    if(x0 >= x1 || len == 0)
+    {
+        if(bbox_)
+            *bbox_ = Rect(org.x, org.y, 0, 0);
+        return org;
+    }
+
+    saved_weights[BUILTIN_FONTS_NUM] = stbtt_GetWeight(fontface->ttface);
+    fontface->setParams(size, weight);
+
+    for(j = 0; j < BUILTIN_FONTS_NUM; j++)
+    {
+        FontFace& fface = builtin_ffaces[j];
+        if(!builtin_ffaces_initialized)
+            fface.set(builtinFontData[j].name);
+
+        if (fface->ttface) {
+            saved_weights[j] = stbtt_GetWeight(fface->ttface);
+            fface->setParams(size, weight);
+        }
+    }
+    builtin_ffaces_initialized = true;
```

Comment:
What is about reenterability?
Do we guarantee avoiding of memory leaks from such initialization?

Reply:
FontEngine is thread-local. There should not be any problems with reenterability.